src/mlpack/methods/ann/gan/gan_impl.hpp

```diff
@@ -370,7 +369,10 @@ template<
 >
 void GAN<Model, InitializationRuleType, Noise, PolicyType>::Shuffle()
 {
-  math::ShuffleData(predictors, responses, predictors, responses);
+  arma::uvec ordering = arma::shuffle(arma::linspace<arma::uvec>(0,
+      numFunctions - 1, numFunctions));
+  arma::mat temp = predictors.cols(ordering);
+  predictors.cols(0, numFunctions - 1) = temp;
```

Comment:
Hmm, can't we do away with the use of temporary variable here? I think it might be a good optimization.

Reply:
I tried but it seems that discriminator predictors remain unchanged if i assign directly. Not sure why.

Reply:
Hmm maybe @rcurtin can help us with the best approach here?

Reply:
I'm surprised that `predictors = predictors.cols(ordering)` doesn't work.  (Also, is `predictors.n_cols == numFunctions`?  If not my code in the last sentence has to be fixed on the left side to `predictors.cols(0, numFunctions - 1)` I guess.)

In any case, I'm also a bit confused on why we aren't shuffling `responses` here too; I think `ShuffleData()` should work.

Another option, assuming that `predictors.n_cols == numFunctions - 1`, is this:

```
arma::mat temp = predictors.cols(ordering);
predictors = std::move(temp);
```

That at least avoids the copy from `temp` to `predictors`, and I think is roughly how Armadillo would run the operation under the hood anyway.

Reply:
There's no need of shuffling responses, because they're all the same (arma::ones). So actually we're saving on computation by not using `ShuffleData()`. Even I think that we should be able to shuffle a few contiguous columns within a range of indices.

Reply:
I was implementing this wrong before which is why it wasn't working. I have removed the use of temporary variable now and `predictors = predictors.cols(ordering)` works as expected.

We are only shuffling the real images here so, that the real images and fake ones produced by the generator remain as distinct groups. The responses for all the real images are same i.e. `1` and therefore do not require shuffling.

Reply:
Understood!  Thanks for the clarification. :+1:  Glad that you were able to get the code working. :)